tracing/span.go

```diff
@@ -306,11 +330,92 @@ func (s *Span) CreateLocalChild(name string, component string) *Span {
 // Parent id will be inherited from the span id,
 // and span id will be randomly generated.
 // Trace id, sampled, and flags will be copied over.
-func (s *Span) CreateLocalChildForContext(ctx context.Context, name string, component string) (context.Context, *Span) {
-	child := s.CreateLocalChild(name, component)
+func CreateLocalChildForContext(ctx context.Context, span Span, name, component string) (context.Context, *LocalSpan) {
+	child := CreateLocalChild(span, name, component)
 	return context.WithValue(ctx, childSpanKey, child), child
 }
 
+// ChildAndThriftContext creates both a client child span and a context can be
+// used by the thrift client code.
+//
+// A thrift client call would look like:
+//
+//     clientCtx, span := parentSpan.ChildAndThriftContext(ctx, "myCall")
+//     result, err := client.MyCall(clientCtx, arg1, arg2)
+//     span.End(ctx, err)
+func ChildAndThriftContext(ctx context.Context, span Span, name string) (context.Context, *ClientSpan) {
+	ctx, child := CreateClientChildForContext(ctx, span, name)
+	ctx = CreateThriftContextFromSpan(ctx, child)
+	return ctx, child
+}
+
+func initChildSpan(parent Span, child Span) {
+	for _, hook := range parent.Hooks() {
+		if err := hook.OnCreateChild(child); err != nil {
+			parent.LogError("OnCreateChild hook error: ", err)
+		}
+	}
+	StartSpan(child)
+}
+
+type baseSpan struct {
+	trace *trace
+	hooks []SpanHook
+
+	name string
+}
+
+// Name returns the name of the Span.
+func (s *baseSpan) Name() string {
```

Comment:
Since `baseSpan` is an unexported type, I think golint won't force you to write doc comments on its exported functions. These comments should be moved to the interface instead.

Reply:
Also `baseSpan` is basically just 3 pointers, so there's no need to really use pointer as the receiver type. Usually we use non-pointer for immutable functions, and pointer for mutable functions, unless the struct is huge or there are some fields not copyable.

Reply:
That's true for everything but `AddHooks`, it appears that `s.hooks` is copied when I don't use a pointer because the Hooks are not actually added to the Span when the receiver is not a pointer.

Reply:
Yes, AddHooks is a mutate function, others are not.

Reply:
See https://godoc.org/github.com/reddit/baseplate.go/batcherror#BatchError for an example, only `Add` and `Clear` uses pointers because these are the only functions that will mutate the state.

Reply:
`SetDebug`, `SetTag` and `AddCounter` are other mutate functions (they don't need pointer receivers to work because you are using pointer fields, but for documentation reasons you should also use pointer receivers for them, but then `baseSpan` is an unexported type so that probably doesn't matter).